streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,252 @@
+        }
+        return assignment;
+    }
+
+    /**
+     * Maps tasks to clients with caught-up states for the task.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return map from tasks with caught-up clients to the list of client candidates
+     */
+    private Map<TaskId, List<ID>> tasksToCaughtUpClients(final SortedMap<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<TaskId, List<ID>> taskToCaughtUpClients = new HashMap<>();
+        for (final SortedMap.Entry<TaskId, SortedSet<ClientIdAndLag<ID>>> taskToRankedClient : statefulTasksToRankedClients.entrySet()) {
+            final SortedSet<ClientIdAndLag<ID>> rankedClients = taskToRankedClient.getValue();
+            for (final ClientIdAndLag<ID> clientIdAndLag : rankedClients) {
+                if (clientIdAndLag.lag() == Task.LATEST_OFFSET || clientIdAndLag.lag() == 0) {
+                    final TaskId taskId = taskToRankedClient.getKey();
+                    taskToCaughtUpClients.computeIfAbsent(taskId, ignored -> new ArrayList<>()).add(clientIdAndLag.clientId());
+                } else {
+                    break;
+                }
+            }
+        }
+        return taskToCaughtUpClients;
+    }
+
+    /**
+     * Maps a task to the client that host the task according to the previous assignment.
+     *
+     * @return map from task IDs to clients hosting the corresponding task
+     */
+    private Map<TaskId, ID> tasksToPreviousClients(final Map<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        final Map<TaskId, ID> tasksToPreviousClients = new HashMap<>();
+        for (final Map.Entry<TaskId, SortedSet<ClientIdAndLag<ID>>> taskToRankedClients : statefulTasksToRankedClients.entrySet()) {
+            final ClientIdAndLag<ID> topRankedClient = taskToRankedClients.getValue().first();
+            if (topRankedClient.lag() == Task.LATEST_OFFSET) {
+                tasksToPreviousClients.put(taskToRankedClients.getKey(), topRankedClient.clientId());
+            }
+        }
+        return tasksToPreviousClients;
+    }
+
+    /**
+     * Assigns task for which one or more caught-up clients exist to one of the caught-up clients.
+     *
+     * @param assignment assignment
+     * @param tasksToCaughtUpClients map from task IDs to lists of caught-up clients
+     */
+    private void assignTasksWithCaughtUpClients(final Map<ID, List<TaskId>> assignment,
+                                                final Map<TaskId, List<ID>> tasksToCaughtUpClients,
+                                                final Map<TaskId, SortedSet<ClientIdAndLag<ID>>> statefulTasksToRankedClients) {
+        // If a task was previously assigned to a client that is caught-up and still exists, give it back to the client
+        final Map<TaskId, ID> tasksToPreviousClients = tasksToPreviousClients(statefulTasksToRankedClients);
```

Comment:
Since this only returns tasks whose previous client was caught up, I think we can simplify the first half of `assignTasksWithCaughtUpClients`  as just
```
tasksToPreviousClients.forEach((t, c) -> assignment.computeIfPresent(c, (k, v) -> { v.add(t);  return v;}));
unassignedTasksWithCaughtUpClients = new ArrayList<>(tasksToCaughtUpClients.keySet());
unassignedTasksWithCaughtUpClients.removeAll(tasksToPreviousClients.keySet());
```